### PR TITLE
Add resetful APIs: set queueupdate flag  

### PR DESCRIPTION
**Pull request by yorma**
1. /servers/:id/queueupdate    --- set queueupdate flag for server
2. /cachegroups/:id/queueupdate  --- set queueupdate flag for all servers in cachegroup.

The APIs document is in https://docs.google.com/document/d/13GjIizusc7iUZq7TuVrsp8JZ2eEx52xUtoX9uRNzFgc/edit?usp=sharing

### CODE REVIEW

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst [hunk 1]
```diff
@@ -332,3 +332,65 @@ Cache Group
...
+
+  **Request Route Parameters**
+
+  +-----------------+----------+----------------------+
+  | Name            | Required | Description          |
+  +=================+==========+======================+
+  | id              | yes      | the cachegroup id.   |
+  +-----------------+----------+----------------------+
+
+  **Request Route Properties**
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: These are just "request properties" not "request route properties". route properties are the things you see in the route.

File: docs/source/development/traffic_ops_api/v12/server.rst [hunk 2]
```diff
@@ -780,3 +780,49 @@ Server
...
+  +===========+==========+==================+
+  | id        | yes      | the server id.   |
+  +-----------+----------+------------------+
+
+  **Request Route Properties**
+
+  +--------------+---------+-----------------------------------------------+
+  | Name         | Type    | Description                                   |
+  +==============+=========+===============================================+
+  | queueUpdate  | integer | The value of queue update flag to set. 0 or 1 |
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/server.rst]: it looks like you take the queueUpdate value and put it in the database. this seems error prone as I can send in the value of 7.

can you change the property name to "action" or "command" that accepts 2 values - queue or dequeue

and then depending on the value, set upd_pending = 0 or 1

...or...

you could create 2 routes:

POST /api/1.2/servers/{:id}/queueupdate
POST /api/1.2/servers/{:id}/dequeueupdate

and get rid of the queueupdate request body parameter

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst [hunk 3]
```diff
@@ -332,3 +332,65 @@ Cache Group
     }
    
 |
+
+**POST /api/1.2/cachegroups/{:id}/queueupdate**
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: should be POST /api/1.2/cache_groups/{:id}/queue_update

File: docs/source/development/traffic_ops_api/v12/server.rst [hunk 4]
```diff
@@ -780,3 +780,49 @@ Server
   +------------------+--------+------------------------------------------------+
   
 |
+
+**POST /api/1.2/servers/{:id}/queueupdate**
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/server.rst]: should be POST /api/1.2/servers/{:id}/queue_update

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst (hunk 3 above)
Comment (yorma) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: ok

File: docs/source/development/traffic_ops_api/v12/server.rst (hunk 4 above)
Comment (yorma) [inline docs/source/development/traffic_ops_api/v12/server.rst]: ok

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst (hunk 1 above)
Comment (yorma) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: ok

File: docs/source/development/traffic_ops_api/v12/server.rst (hunk 2 above)
Comment (yorma) [inline docs/source/development/traffic_ops_api/v12/server.rst]: I prefer your 1st suggestion, will change the property name to "action" that accepts "queue" or "dequeue". Same change will apply to upper API /api/1.2/cachegroups/{:id}/queue_update .

**Comment by yorma** [comment]
@mitchell852 changed according to your comments, please review. thanks

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst (hunk 3 above)
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: yes, i agree.

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst [hunk 5]
```diff
@@ -332,3 +332,65 @@ Cache Group
...
+  +-----------------+---------+----------------------------------------------------+
+  | serverNames     | array   | servers name array in the cachegroup in cdn        |
+  +-----------------+---------+----------------------------------------------------+
+
+  **Response Example** ::
+
+    {
+      "response": {
+            "cachegroupName": "us-il-chicago",
+            "udpPending": "1",
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: udpPending should be updPending

or do you want to echo back the action instead?

"action": "queue"
"action": "dequeue"

File: docs/source/development/traffic_ops_api/v12/server.rst [hunk 6]
```diff
@@ -780,3 +780,49 @@ Server
...
+  +--------------+---------+-----------------------------------------------+
+  | serverId     | integer | server id                                     |
+  +--------------+---------+-----------------------------------------------+
+
+  **Response Example** ::
+
+    {
+      "response": {
+          "serverId": "1",
+          "udpPending": "1"
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/server.rst]: same as comment in cachegroup.rst

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst (hunk 5 above)
Comment (yorma) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: ok, will change echo back action
"action": "queue"
"action": "dequeue"
thanks,

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst [hunk 7]
```diff
@@ -332,3 +332,65 @@ Cache Group
     }
    
 |
+
+**POST /api/1.2/cachegroups/{:id}/queue_update**
+
+  Set server queue update flag value for all servers in the cachegroup and cdn.
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: Can you change this to something like:

"Queue or dequeue updates for all servers assigned to a cache group limited to a specific CDN"

File: docs/source/development/traffic_ops_api/v12/server.rst [hunk 8]
```diff
@@ -780,3 +780,49 @@ Server
   +------------------+--------+------------------------------------------------+
   
 |
+
+**POST /api/1.2/servers/{:id}/queue_update**
+
+  Set server queue update flag value.
```
Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/server.rst]: Can you change this to:

"Queue or dequeue updates for a specific server"

File: traffic_ops/app/lib/API/Cachegroup.pm [hunk 9]
```diff
@@ -240,5 +240,75 @@ sub get_typeId {
     }
     return($type_id);
 }
+sub postupdatequeue {
+	my $self       = shift;
+    my $params = $self->req->json;
+	if ( !&is_oper($self) ) {
+        return $self->forbidden();
+    }
+
```
Comment (dangogh) [inline traffic_ops/app/lib/API/Cachegroup.pm]: lots of places where the indentation is incorrect..     Please run perltidy with the options file from the dev docs.

File: traffic_ops/app/lib/API/Cachegroup.pm [hunk 10]
```diff
@@ -240,5 +240,75 @@ sub get_typeId {
...
+    my $setqueue = $params->{action};
+	if ( !defined($setqueue)) {
+        return $self->alert("action needed, should be queue or dequeue.");
+    }
+	if ( $setqueue eq "queue") {
+		$setqueue = 1
+	} else {
+		if ($setqueue eq "dequeue") {
+			$setqueue = 0
```
Comment (dangogh) [inline traffic_ops/app/lib/API/Cachegroup.pm]: elsif?

File: docs/source/development/traffic_ops_api/v12/cachegroup.rst (hunk 7 above)
Comment (yorma) [inline docs/source/development/traffic_ops_api/v12/cachegroup.rst]: ok, thanks

File: traffic_ops/app/lib/API/Cachegroup.pm (hunk 10 above)
Comment (yorma) [inline traffic_ops/app/lib/API/Cachegroup.pm]: yes, thanks

**Comment by yorma** [comment]
@mitchell852 @dangogh , have make modification according to your comments, please review, thanks.

**Comment by mitchell852** [comment]
I'm good with this. I'll let @dangogh do his thing and merge upon his approval.

File: docs/source/development/traffic_ops_api/v12/server.rst [hunk 11]
```diff
@@ -780,3 +780,49 @@ Server
...
+  +===========+==========+==================+
+  | id        | yes      | the server id.   |
+  +-----------+----------+------------------+
+
+  **Request Properties**
+
+  +--------------+---------+-----------------------------------------------+
+  | Name         | Type    | Description                                   |
+  +==============+=========+===============================================+
+  | action       | string  | queue or dequeue the udp_pending              |
```
Comment (dangogh) [inline docs/source/development/traffic_ops_api/v12/server.rst]: missed this one -- udp... -> upd...

Comment (mitchell852) [inline docs/source/development/traffic_ops_api/v12/server.rst]: i'd say change the description to "queue or dequeue"

**Comment by yorma** [comment]
@mitchell852 @dangogh , I remove udp_pending according your comments. thanks.

**Pull request merged**